### PR TITLE
[JENKINS-67664][JENKINS-59652] Add retry mechanism for container exec…

### PR DESCRIPTION
**Pull request by Dohbedoh**
… websocket connections

Proposing a retry mechanism to mitigate sporadic issues as encountered in [JENKINS-67664](https://issues.jenkins.io/browse/JENKINS-67664) and [JENKINS-59652](https://issues.jenkins.io/browse/JENKINS-59652). as the mechanism of the decorator can be fragile depending on environments. Also amends https://github.com/jenkinsci/kubernetes-plugin/pull/1159.

If the websocket connection is unsuccessful for any reason, the launch reattempt it after `3` seconds. After `3` failed attempts it gives up.

* ~The retry wait time of 3 seconds is configurable with `org.csanchez.jenkins.plugins.kubernetes.pipeline.ContainerExecDecorator.websocketConnectionRetryWait`. Default to `3` seconds.~
* ~The number of retry is configurable with `org.csanchez.jenkins.plugins.kubernetes.pipeline.ContainerExecDecorator.websocketConnectionMaxRetries`. Default to `3` times.~
* The number of retry is configurable with `org.csanchez.jenkins.plugins.kubernetes.pipeline.ContainerExecDecorator.websocketConnectionMaxRetry`. Default to `5` times.
* The maximum backoff time between retries is configurable with `org.csanchez.jenkins.plugins.kubernetes.pipeline.ContainerExecDecorator.websocketConnectionMaxRetryBackoff`. Default to `30` seconds.

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Comment by Dohbedoh** [comment]
Added exp backoff. With 5 max retries and 30 max backoff time in seconds by default. I need to test this in a test environment first though.

**Comment by jglick** [comment]
Rather than wasting time trying to fix the current impl of this decorator, which is known to be fundamentally not scalable due to use of the API server, why not actually fix it to stream commands and (non-durable-task) responses over the Remoting channel as has been long proposed?

**Comment by Dohbedoh** [comment]
@jglick it feels like a quick workaround that could have great impact and help users until we can design an alternative solution.

**Comment by Dohbedoh** [comment]
It mostly is. I put the snippet that handle the websocket connection (opening) in a function `openExecInContainer`.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java [hunk 1]
```diff
@@ -382,77 +385,42 @@ private Proc doLaunch(boolean quiet, String[] cmdEnvs, OutputStream outputForCal
...
-                                e);
-                    } else {
-                        throw e;
+                int attempts = 0;
+                ExecWatch watch = null;
+                AtomicBoolean alive = null;
+                CountDownLatch finished = null;
+                while (attempts < WEBSOCKET_CONNECTION_MAX_RETRY) {
+                    try {
+                        alive = new AtomicBoolean(false);
```
Comment (Dohbedoh) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java:394]: I do have doubt about `alive` and `finished` not being truly final anymore. I wonder if there is a possibility that those references be set by a previous attempts (some kind of late `onClose` / `onFailure`). Maybe we should create an ExecWatchWrapper that hold reference to the watch and the `alive` / `finished` associated. Might be safer...

**Comment by Dohbedoh** [comment]
I put back the all snippet where it was, also did some improvement. Hopefully that is simpler to review as is.

**Comment by SeriousMatt** [comment]
@jglick can this review be expedited? we've been running out of options for using the k8s plugin reliably due to the 5000 millisecond error in Jenkins

**Comment by jglick** [comment]
@SeriousMatt I am not a maintainer and am not reviewing.

There is a simple workaround: run single-container pods and do not use the `container` step.

**Comment by SeriousMatt** [comment]
thanks, we use the jnlp and a separate deploy container in a pod. combining into a single container is not unfortunately a simple process for our setup but we'll give it a try

**Comment by brudnyhenry** [comment]
Hi @jglick this issue is forcing us to migrate a couple of projects from Jenkins to Gitlab CI. Jenkins stability is unacceptable - jobs failing randomly due to this error, please prioritize it if you can. I hope you were joking about the single container approach- all major CI/CD tools support running jobs in multiple containers, which is really the benefit of having a dynamic agent in K8. With this approach way, we would soon recommend spinning up bare metal agents

**Comment by Dohbedoh** [comment]
The retry mechanism seems to have helped according to [this comment](https://issues.jenkins.io/browse/JENKINS-67664?focusedCommentId=428631&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-428631). @jenkinsci/kubernetes-plugin-developers any chance to get further review and help make some progress here ?

**Comment by jglick** [comment]
> I hope you were joking about the single container approach

As I said, it is a workaround, since the design of the `container` step has long been known to be flawed. A single-container agent pod has better performance, with the obvious drawback that you are forced to prepare a container image with all required tooling as well as the Jenkins agent and its JRE.

> any chance to get further review

AFAIK the maintainer of this plugin is unavailable for a couple of weeks.

**Comment by jglick** [comment]
BTW @Dohbedoh I merged in `master` to make sure this gets an incremental deployment (assuming tests pass). Please refer to that URL in the `incrementals` repo (you will get a check here) when asking users to test prerelease versions, rather than file attachments or downloads from ci.jenkins.io.

**Review by amuniz — APPROVED** [review]
Looks good to me.

**Review by twasyl — APPROVED** [review]
Good job

**Review by mikecirioli — APPROVED** [review]
The retry mechanism looks solid to me, i am only moderately familiar with this plugin overall though.

**Review by carlossg — APPROVED** [review]

**Comment by jmhardison** [comment]
Adding a note; we have deployed the incremental build into our environment. We normally have somewhere close to 100 k8s agent jobs per hour that cycle through our env during business hours.

Probably too early for me to say unequivocally it's helped, but we have changed our configuration back to a k8s idle timeout of 0 so agents are created and destroyed more often. Last we had this configuration it increased the observed rate of `5000ms` errors. 

So far, no occurrences of the error have been observed yet. We will report back in a few days with more info if this is still open.

**Comment by Vlatombe** [comment]
> There is a simple workaround: run single-container pods and do not use the container step.

To elaborate on this: you CAN use multiple containers pods, such as these use cases where you want a service next to the agent pod (a database, a docker server, etc.), however, the `container` step has design issues as Jesse mentioned, so as long as you don't use it to execute any interactive command to a side container, you're safe from this particular issue.

> forcing us to migrate a couple of projects from Jenkins to Gitlab CI

As far as I know, Gitlab doesn't have anything equivalent to the `container` step. So whatever you're doing is based on misinterpretation or misusage of either the Jenkins Kubernetes plugin or Gitlab.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java [hunk 2]
```diff
@@ -382,78 +386,115 @@ private Proc doLaunch(boolean quiet, String[] cmdEnvs, OutputStream outputForCal
...
                                             "onClose called but latch already finished. This indicates a bug in the kubernetes-plugin");
+                                    }
+                                    finished.countDown();
                                 }
-                                finished.countDown();
+                            }).exec(sh);
+                        
+                        // prevent a wait forever if the connection is closed as the listener would never be called
+                        try {
+                            if(started.await(WEBSOCKET_CONNECTION_TIMEOUT, TimeUnit.SECONDS)) {
```
Comment (Vlatombe) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java:455]:
```suggestion
                            if (started.await(WEBSOCKET_CONNECTION_TIMEOUT, TimeUnit.SECONDS)) {
```

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java [hunk 3]
```diff
@@ -382,78 +386,115 @@ private Proc doLaunch(boolean quiet, String[] cmdEnvs, OutputStream outputForCal
...
+                        } catch (InterruptedException e) {
+                            closeWatch(watch);
+                            throw e;
+                        }
+                        
+                    } catch (KubernetesClientException e) {
+                        launcher.getListener().getLogger().print("Failed to start websocket connection: ");
+                        
+                        // In case of 400 / Bad Request, do not attempt a retry
+                        if(e.getCause() instanceof WebSocketHandshakeException) {
```
Comment (Vlatombe) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java:473]:
```suggestion
                        if (e.getCause() instanceof WebSocketHandshakeException) {
```

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java [hunk 4]
```diff
@@ -382,78 +386,115 @@ private Proc doLaunch(boolean quiet, String[] cmdEnvs, OutputStream outputForCal
...
+                            WebSocketHandshakeException wsException = (WebSocketHandshakeException) e.getCause();
+                            if(wsException.getResponse() != null && wsException.getResponse().code() == 400) {
```
Comment (Vlatombe) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java:475]:
```suggestion
                            if (wsException.getResponse() != null && wsException.getResponse().code() == 400) {
```

**Review by Vlatombe — APPROVED** [review]

**Comment by jglick** [comment]
> As far as I know, Gitlab doesn't have anything equivalent to the `container` step

@Dohbedoh noted https://gitlab.com/gitlab-org/gitlab-runner/-/merge_requests/1775 which sounds potentially analogous.

**Comment by s7an-it** [comment]
What I don't understand is how one month ago I was able to run multiple containers with shell executions and now no matter to what I migrate I simply get the 500 internal server websocket error very frequently if use anything outside of the inbound, retry on kuberentes doesn't work and I can't influence the timeout or anything.
Something went wrong somewhere, maybe there is a problem with EKS 1.22 specifically, but for certain on EKS.1.20 with old jnlp containers and pre 2.36x Jenkins was a powerhouse in comparison to now, simply because it's hyper unstable and lacks multi container with exec.

**Pull request merged**